### ISSUE TITLE
Make it possible to update the website with new documentation

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2380]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2380|
|Type|task|
|Created|2007-01-10 03:58:37Z|
|Branch|https://github.com/twisted/twisted/tree/build-docs-2380-2|

The lore-processed howto documentation is presented on the website.  Updating this documentation is extremely time consuming.

The release process and the website should be organized such that there is a simple command which will cause the new documentation to appear on the website.

Attachments:

* [build-docs-2380.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/382/3820eccabc46f2c4f038c662d6542e8a9bd0de20/0ed5e5e956532bd5c002cb7b183c2c6bc3a8457b.patch) (5820 bytes) - added by thijs on 2012-02-09 10:59:48Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2380 2380
type__task task
reporter__exarkun exarkun
priority__normal normal
milestone__totally_automated_release_infrastructure totally automated release infrastructure
branch__branches_build_docs_2380_2 branches/build-docs-2380-2
branch_author__thijs__therve thijs, therve
status__closed closed
resolution__fixed fixed
component__release_management release management
keywords__ 
time__1168401517000000 1168401517000000
changetime__1363798186000000 1363798186000000
version__None None
owner__therve therve
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

[#4355](https://github.com/twisted/twisted/issues/4355) was a duplicate of this.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

(In [33530]) Branching to 'build-docs-2380'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

(In [33535]) add build-docs script, refs [#2380](https://github.com/twisted/twisted/issues/2380)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

I copied the temporary script on [ReleaseProcess#Buildhowtodocumentsforwebsite] and hooked it up to a `build-docs` script. 

```
$ bin/admin/build-docs
Must specify three arguments: Twisted checkout path, destination path, and template path.
$ bin/admin/build-docs . ./book ../../../../../twisted-website/docs/website-template.tpl 
pnmtops: writing color PostScript...
pnmtops: writing color PostScript...
pnmtops: writing color PostScript...
pnmtops: writing color PostScript...
pnmtops: writing color PostScript...
pnmtops: writing color PostScript...
pnmtops: warning, image too large for page, rescaling to 0.93578
pnmtops: writing color PostScript...
ls -lh book/
total 1.4M
-rwxrwxr-x 1 thijs thijs 1.4M 2012-02-10 18:14 book.pdf
```

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_8"></a>|@jonathanj set owner to @thijstriemstra|
|-|-|

Thanks for doing this.

My main concern here is the lack of test coverage for the behaviour of `BuildDocsScript.buildDocs`. I think breaking `buildDocs` into its smaller components and parameterizing the builder type would make the code both easier to test and read.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

(In [37718]) Branching to 'build-docs-2380-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve removed owner|
|-|-|

Alright, there is one test now, plus a couple of other fixes. Please review!

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_11"></a>|@jonathanj set owner to @therve|
|-|-|

Thanks for coming back to this!

I think the only item is that `setupTeXFiles`, `_setupTeXSections` and `_setupTeXBook` need docstrings.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set status to `closed`|
|-|-|

(In [37746]) Merge build-docs-2380-2

Authors: thijs, therve
Reviewer: jonathanj
Fixes: [#2380](https://github.com/twisted/twisted/issues/2380)

Add a script to build documentation for the release.
